### PR TITLE
[DRAFT] feat(hooks): add onListen hook

### PR DESCRIPTION
**Pull request by Yberion**
#### Checklist

- [ ] run `npm run test` and `npm run benchmark`
- [ ] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

The purpose of this PR is to add an `onListen` application hook which will act similar to `fastify.server.on`.

Following https://github.com/fastify/fastify/issues/4542 I made this PR as a draft.

As said in the issue, I'm not enough experienced with the JS Fastify base code and unit test standard (tap).

I would enjoy to get your feedback and help.

### CODE REVIEW

File: lib/server.js [hunk 1]
```diff
@@ -68,6 +68,11 @@ function createServer (options, httpHandler) {
 
         multipleBindings.call(this, server, httpHandler, options, listenOptions, () => {
           this[kState].listening = true
+
+          for (const hook of this[kHooks].onListen) {
+            hook.call(this)
```
Comment (Eomm) [inline lib/server.js:73]: We need to deal with errors as suggested, but beforehand we need to define how to execute the `onListen` function - serially or parallels.

I would expect:

- all the hooks run serially
- all the async hooks are awaited
- if an error is triggered, the error is logged as a warning

_(before implementing it, let discuss it a bit)_

```suggestion
     try {
        const ret = hook.call(server)
        if (ret && typeof ret.then === 'function') {
          ret.then(done, done)
          return
        }
      } catch (error) {
        err = error
      }
```

Comment (Yberion) [reply lib/server.js:73]: The behavior you described looks fine to me.

I'm not sure this is the best place to put both check for this new hook.

I also didn't know how to make it work with `done`.

Comment (Yberion) [reply lib/server.js:73]: Hello @Eomm, maybe I misunderstood, do you require anything else from me? I will enjoy to make this feature getting forward with your help. 😃

File: lib/server.js [hunk 2]
```diff
@@ -86,6 +91,11 @@ function createServer (options, httpHandler) {
           return new Promise((resolve, reject) => {
             multipleBindings.call(this, server, httpHandler, options, listenOptions, () => {
               this[kState].listening = true
+
+              for (const hook of this[kHooks].onListen) {
+                hook.call(this)
+              }
```
Comment (mcollina) [inline lib/server.js:97]: This should use our hooks system

File: lib/server.js (hunk 1 above)
Comment (Eomm) [reply lib/server.js:73]: I would like feedback from @fastify/core about my comment above 👍🏼

Comment (Uzlopak) [reply lib/server.js:73]: I expect the same.

**Comment by metcoder95** [comment]
I’ll take a look into this one again soon :)

**Comment by Yberion** [comment]
Thanks, as I said I'm up to follow your advices since I'm not familiar enough with the JS codebase 😃

**Comment by metcoder95** [comment]
Feel free to open a new draft PR and we can pair together from there :)

**Pull request closed without merging**